[FLINK-32416] initial implementation of DynamicKafkaSource with bound…

mas-chen opened the pull request:
…ed/unbounded support and unit/integration tests

### What is the purpose of the change
[FLIP-246](https://cwiki.apache.org/confluence/x/CBn1D) describes the changes and this PR follows the design illustrated in the doc. 

### Brief Change Log
* New Public API: DynamicKafkaSource, DynamicKafkaSourceBuilder, KafkaMetadataService and related classes, etc.
* Supports multi cluster read, dynamic cluster/topic read
* Supports multi cluster, dynamic metrics
* Supports StoppableKafkaEnumContextProxy to workaround and provide single threaded guarantees
* Supports unbounded and bounded mode
* Only supports configurations/deserializers defined uniformly across clusters, can be extended later on
* 3 stream subscription methods
* 1 KafkaMetadataService that supports single cluster, can expose more implementations in public API depending on discussion
* Supports metadata discovery interval as well as error handling strategy
* Add support in test utils to bring up 2 kafka clusters for testing (tried to make least changes as possible)

### Verifying this change
* Unit Tests
* Integration Tests (including covering migration scenarios)
* Connector Test Framework Integration

mxm commented:
Hey Mason! Do you want a review for this?

mas-chen commented:
@mxm Yes please! I've rebased the PR and working on fixing some flaky unit tests

mxm commented on flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/DynamicKafkaSource.java:
```diff
@@ -0,0 +1,199 @@
+ * href="https://cwiki.apache.org/confluence/x/CBn1D">FLIP-246: DynamicKafkaSource</a>
+ *
+ * <p>This source's key difference from {@link KafkaSource} is that it enables users to read
+ * dynamically from topics, which belong to one or more clusters.
```
We should clarify what "dynamically" means.

mxm reviewed (commented):
Thanks @mas-chen! The changes look very solid to me and they capture what has been proposed in the FLIP. The only thing we are missing here is documentation. It doesn't have to be much but we should at least outline some basic usage examples.

tzulitai reviewed (approved):
@mas-chen I've been testing this with a real, simple implementation of the KafkaMetadataService over the winter break and it has worked nicely, matching what is described in the FLIP. Added test time to the build is also acceptable. The code overall looks pretty solid as well.

Since this new connector is marked as `@Experimental`, I think we're in a good state to merge it to be included for Kafka connector 3.1.0 release. What do you think @mas-chen?

mas-chen commented on flink-connector-kafka/src/test/resources/log4j2-test.properties:
```diff
@@ -36,3 +36,12 @@ logger.zookeeper.name = org.apache.zookeeper
 logger.I0Itec.name = org.I0Itec
 logger.I0Itec.level = OFF
+
+logger.DynamicKafkaSourceReader.name = org.apache.flink.connector.kafka.source.reader.DynamicKafkaSourceReader
+logger.DynamicKafkaSourceReader.level = OFF
+
+logger.kafkaSourceReader.name = org.apache.flink.connector.kafka.source.reader.KafkaSourceReader
+logger.kafkaSourceReader.level = OFF
+
+logger.SourceReaderBase.name = org.apache.flink.connector.base.source.reader.SourceReaderBase
+logger.SourceReaderBase.level = OFF
```
This was for troubleshooting integration tests. I'll remove them now that they are stable.

mas-chen commented on flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/DynamicKafkaSource.java:
```diff
@@ -0,0 +1,199 @@
+ * @param <T> Record type
+ */
+@PublicEvolving
+public class DynamicKafkaSource<T>
```
Yes, I was a bit lazy there, since it required the diagrams to change. Let me revisit that now

mas-chen commented:
@tzulitai Thanks for the review. I would like get some more input from at least another reviewer from merging this and I agree this should be included in the 3.1.0 release. Perhaps @mxm?

After this PR, I still need to complete the documentation for this new connector.

mas-chen commented on flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/enumerator/KafkaSourceEnumState.java:
```diff
@@ -42,7 +42,7 @@ public KafkaSourceEnumState(
     }
 
-    KafkaSourceEnumState(
+    public KafkaSourceEnumState(
```
This is the only KafkaSource code that I had to modify. As a result of moving packages, this needed to be publicly visible (the class is still annotated as Internal)

mxm commented on flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/DynamicKafkaSource.java:
```diff
@@ -0,0 +1,199 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.kafka.source;
```
On hindsight, I wonder whether `source.dynamic` makes more sense than `dynamic.source` 🙈

mas-chen replied:
I actually prefer `dynamic.source` since it makes sense for future usecases to leverage the dynamic code. e.g. a `dynamic.sink`. With that being said, I'm going to move the metadata up one package since that isn't source specific.

mas-chen replied:
Let me know if you have any other considerations. I'll squash the commits if we are in agreement (leaving the dangling commit now for your reviewing convenience)

mas-chen commented:
Thanks @tzulitai @mxm for your review! I've addressed all the comments and squashed the commits. Can you please merge it at your convenience?

I'll start on the docs immediately and hope to use this connector directly in the 3.1.0 release!

mxm commented:
Sounds good. Can we get a bare-minimum version of the documentation into this PR?

mas-chen commented:
Are the Javadocs sufficient in the bare minimum case? Otherwise, I'll just write the documentation into this PR

mxm commented:
A brief one-pager of the feature would be good. We can expand on it later. Just to have something in the web docs.

mas-chen commented:
one-pager docs are here: docs/content/docs/connectors/datastream/dynamic-kafka.md. Marked with experimental status.

Also copied this into the chinese doc directory.

mxm commented:
Awesome, thanks a lot! Merging once tests pass.

MartijnVisser commented:
NB: after this was merged, the CI doesn't work for `main` anymore CC @mas-chen @mxm

mxm commented:
We also noticed that. @mas-chen is looking into fixing this.

mxm commented:
Debug info:
```
2024-01-12T16:49:32.5164583Z [ERROR] Failures: 
2024-01-12T16:49:32.5166818Z [ERROR]   DynamicKafkaSourceITTest$IntegrationTests>SourceTestSuiteBase.testMultipleSplits:211->SourceTestSuiteBase.checkResultWithSemantic:743 Expected to have exactly 40 records in result, but only received 30 records
```

mas-chen commented:
Sorry about that. ^The linked PR should fix this
